src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -0,0 +1,2984 @@
+            var module = (PEModuleSymbol)comp.GetReferencedAssemblySymbol(refB).Modules[0];
+            // Read from metadata directly to inspect the embedded type.
+            var decoder = new MetadataDecoder(module);
+            var reader = module.Module.MetadataReader;
+            var fieldHandle = reader.FieldDefinitions.Single(handle => reader.GetString(reader.GetFieldDefinition(handle).Name) == "F");
+            var fieldInfo = decoder.DecodeFieldSignature(fieldHandle);
+            Assert.True(fieldInfo.IsByRef);
+            Assert.Equal(new[] { "System.Int32", "System.Object" }, fieldInfo.RefCustomModifiers.SelectAsArray(m => m.Modifier.ToTestDisplayString()));
+        }
+
+        private static void VerifyFieldSymbol(FieldSymbol field, string expectedDisplayString, RefKind expectedRefKind, string[] expectedRefCustomModifiers)
+        {
+            Assert.Equal(expectedRefKind, field.RefKind);
+            Assert.Equal(expectedRefCustomModifiers, field.RefCustomModifiers.SelectAsArray(m => m.Modifier.ToTestDisplayString()));
+            Assert.Equal(expectedDisplayString, field.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void DefiniteAssignment_01()
+        {
+            var source =
+@"ref struct S1<T>
+{
+    public ref T F;
+}
+ref struct S2<T>
+{
+    public ref T F;
+    public S2(ref T t) { }
+}
+ref struct S3<T>
+{
+    public ref T F;
+    public S3(ref T t) : this() { }
+}
+ref struct S4<T>
+{
+    public ref T F;
+    public S4(ref T t)
+    {
+        this = default;
+    }
+}
+class Program
+{
+    static void Main()
+    {
+        int i = 0;
+        new S1<int>().F = ref i;
+        new S2<int>().F = ref i;
+        new S3<int>().F = ref i;
+        new S4<int>().F = ref i;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (8,12): error CS0171: Field 'S2<T>.F' must be fully assigned before control is returned to the caller
+                //     public S2(ref T t) { }
+                Diagnostic(ErrorCode.ERR_UnassignedThis, "S2").WithArguments("S2<T>.F").WithLocation(8, 12));
+        }
+
+        [Fact]
+        public void DefiniteAssignment_02()
+        {
+            // Should we report a warning when assigning a value rather than a ref in the
+            // constructor, because a NullReferenceException will be thrown at runtime?
+            var source =
+@"ref struct S<T>
+{
+    public ref T F;
+    public S(ref T t)
+    {
+        F = t;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void AssignValueTo_InstanceMethod_RefField()
+        {
+            var source =
+@"ref struct S<T>
+{
+    public ref T F;
+    public S(T tValue, ref T tRef, out T tOut, in T tIn)
+    {
+        tOut = default;
+        F = tValue;
+        F = tRef;
+        F = tOut;
+        F = tIn;
```

Comment:
Shouldn't all these assignments generate diagnostics? As written they are going to cause a `NullReferenceException`. The definite assignment should be looking at reassignment, not value assignment right?

Reply:
> Shouldn't all these assignments generate diagnostics? As written they are going to cause a `NullReferenceException`. 

Yes, I think assigning a value to a ref field before initializing the ref should be an error reported from definite assignment. The test plan contains an LDM item for this.

Reply:
Agree that we should come up with some kind of guardrail here. We could consider excluding ref fields from "auto-default", with the rationale that we think the "null `ref`" is a deeply unreasonable/unsafe value to use for a ref field.